File: NewENest/go-server/models/study_room.go
package models

import (
	"time"
)

// StudyRoom 自习室模型
type StudyRoom struct {
	ID              int       `json:"id" db:"id"`
	OwnerID         int       `json:"owner_id" db:"owner_id"`
	Name            string    `json:"name" db:"name"`
	Description     string    `json:"description,omitempty" db:"description"`
	ShareLink       string    `json:"share_link,omitempty" db:"share_link"`
	MaxMembers      int       `json:"max_members" db:"max_members"`
	IsPrivate       bool      `json:"is_private" db:"is_private"`
	Theme           string    `json:"theme,omitempty" db:"theme"`
	BackgroundImage string    `json:"background_image,omitempty" db:"background_image"`
	CreatedAt       time.Time `json:"created_at" db:"created_at"`
	ExpiresAt       time.Time `json:"expires_at" db:"expires_at"`
	
	// 关联信息
	Owner           *User     `json:"owner,omitempty" db:"-"`
	Members         []RoomMember `json:"members,omitempty" db:"-"`
	MemberCount     int       `json:"member_count,omitempty" db:"-"`
	OnlineCount     int       `json:"online_count,omitempty" db:"-"`
}

// RoomMember 自习室成员模型
type RoomMember struct {
	ID          int       `json:"id" db:"id"`
	RoomID      int       `json:"room_id" db:"room_id"`
	UserID      int       `json:"user_id" db:"user_id"`
	IsAnonymous bool      `json:"is_anonymous" db:"is_anonymous"`
	Role        string    `json:"role" db:"role"` // 角色：owner, admin, member
	Status      string    `json:"status" db:"status"` // 状态：online, offline, away
	JoinedAt    time.Time `json:"joined_at" db:"joined_at"`
	
	// 关联信息
	User        *User     `json:"user,omitempty" db:"-"`
}

// StudyRoomRepository 自习室仓库接口
type StudyRoomRepository interface {
	FindByID(id int) (*StudyRoom, error)
	FindByShareLink(shareLink string) (*StudyRoom, error)
	GetUserRooms(userID int) ([]StudyRoom, error)
	GetPublicRooms(limit, offset int) ([]StudyRoom, int, error)
	Create(room *StudyRoom) error
	Update(room *StudyRoom) error
	Delete(id int) error
	
	// 成员管理
	GetRoomMembers(roomID int) ([]RoomMember, error)
	GetRoomMember(roomID, userID int) (*RoomMember, error)
	AddMember(member *RoomMember) error
	UpdateMemberStatus(roomID, userID int, status string) error
	RemoveMember(roomID, userID int) error
	
	// 统计信息
	GetRoomMemberCount(roomID int) (int, error)
	GetRoomOnlineCount(roomID int) (int, error)
}

// StudyRoomCreateDTO 创建自习室DTO
type StudyRoomCreateDTO struct {
	Name            string    `json:"name" validate:"required,min=3,max=100"`
	Description     string    `json:"description,omitempty"`
	MaxMembers      int       `json:"max_members" validate:"required,min=1,max=100"`
	IsPrivate       bool      `json:"is_private"`
	Theme           string    `json:"theme,omitempty"`
	BackgroundImage string    `json:"background_image,omitempty"`
	Duration        int       `json:"duration" validate:"required,min=1,max=1440"` // 持续时间（分钟）
}

// StudyRoomUpdateDTO 更新自习室DTO
type StudyRoomUpdateDTO struct {
	Name            string    `json:"name,omitempty" validate:"omitempty,min=3,max=100"`
	Description     string    `json:"description,omitempty"`
	MaxMembers      int       `json:"max_members,omitempty" validate:"omitempty,min=1,max=100"`
	IsPrivate       *bool     `json:"is_private,omitempty"`
	Theme           string    `json:"theme,omitempty"`
	BackgroundImage string    `json:"background_image,omitempty"`
}

// StudyRoomDTO 自习室信息DTO
type StudyRoomDTO struct {
	ID              int       `json:"id"`
	Name            string    `json:"name"`
	Description     string    `json:"description,omitempty"`
	ShareLink       string    `json:"share_link,omitempty"`
	MaxMembers      int       `json:"max_members"`
	IsPrivate       bool      `json:"is_private"`
	Theme           string    `json:"theme,omitempty"`
	BackgroundImage string    `json:"background_image,omitempty"`
	CreatedAt       time.Time `json:"created_at"`
	ExpiresAt       time.Time `json:"expires_at"`
	Owner           *UserProfileDTO `json:"owner,omitempty"`
	MemberCount     int       `json:"member_count"`
	OnlineCount     int       `json:"online_count"`
}

// ToRoomDTO 将StudyRoom转换为StudyRoomDTO
func (r *StudyRoom) ToRoomDTO() *StudyRoomDTO {
	if r == nil {
		return nil
	}
	
	return &StudyRoomDTO{
		ID:              r.ID,
		Name:            r.Name,
		Description:     r.Description,
		ShareLink:       r.ShareLink,
		MaxMembers:      r.MaxMembers,
		IsPrivate:       r.IsPrivate,
		Theme:           r.Theme,
		BackgroundImage: r.BackgroundImage,
		CreatedAt:       r.CreatedAt,
		ExpiresAt:       r.ExpiresAt,
		Owner:           r.Owner.ToProfileDTO(),
		MemberCount:     r.MemberCount,
		OnlineCount:     r.OnlineCount,
	}
}

// RoomMemberDTO 自习室成员DTO
type RoomMemberDTO struct {
	ID          int       `json:"id"`
	UserID      int       `json:"user_id"`
	Username    string    `json:"username"`
	Avatar      string    `json:"avatar,omitempty"`
	IsAnonymous bool      `json:"is_anonymous"`
	Role        string    `json:"role"`
	Status      string    `json:"status"`
	JoinedAt    time.Time `json:"joined_at"`
}

// ToMemberDTO 将RoomMember转换为RoomMemberDTO
func (m *RoomMember) ToMemberDTO() *RoomMemberDTO {
	if m == nil || m.User == nil {
		return nil
	}
	
	return &RoomMemberDTO{
		ID:          m.ID,
		UserID:      m.UserID,
		Username:    m.User.Username,
		Avatar:      m.User.Avatar,
		IsAnonymous: m.IsAnonymous,
		Role:        m.Role,
		Status:      m.Status,
		JoinedAt:    m.JoinedAt,
	}
} 

File: NewENest/web-app/src/api/studyRoom.ts
import axios from 'axios';
import type { AxiosResponse } from 'axios';
import { useUserStore } from '@/stores/user';
import api from '@/config/api';

// API响应类型定义
interface ApiResponse<T> {
  code: number;
  message: string;
  data?: T;
  errors?: string[] | string;
}

// 自习室相关DTO类型定义
export interface StudyRoomDTO {
  id: number;
  name: string;
  description: string;
  share_link?: string;
  max_members: number;
  is_private: boolean;
  theme: string;
  background_image?: string;
  created_at: string;
  expires_at: string;
  owner?: {
    id: number;
    username: string;
    avatar?: string;
  };
  member_count: number;
}

export interface StudyRoomDetailDTO extends StudyRoomDTO {
  members?: RoomMemberDTO[];
}

export interface RoomMemberDTO {
  user_id: number;
  username?: string;
  avatar?: string;
  is_anonymous: boolean;
  role: string;
  status: string;
  joined_at: string;
}

export interface CreateStudyRoomDTO {
  name: string;
  description: string;
  max_members: number;
  is_private: boolean;
  theme?: string;
  background_image?: string;
  expires_in: number; // 过期时间（小时）
}

export interface UpdateStudyRoomDTO {
  name?: string;
  description?: string;
  max_members?: number;
  is_private?: boolean;
  theme?: string;
  background_image?: string;
}

// 使用从config/api.ts导入的全局api实例，删除这里的重复定义
// const api = axios.create({
//   baseURL: import.meta.env.VITE_API_URL || 'http://localhost:8080/api',
//   timeout: 10000,
//   headers: {
//     'Content-Type': 'application/json',
//     'Accept': 'application/json',
//   }
// });

// 请求和响应拦截器已在config/api.ts中定义，此处无需重复
// 删除拦截器代码
// ...

export const studyRoomApi = {
  // 获取自习室列表
  getStudyRooms: async (page: number = 1, pageSize: number = 10): Promise<{rooms: StudyRoomDTO[], total: number}> => {
    const response = await api.get('/study-rooms', { params: { page, pageSize } });
    return response.data || { rooms: [], total: 0 };
  },
  
  // 获取公开自习室列表
  getPublicStudyRooms: async (page: number = 1, pageSize: number = 10): Promise<{rooms: StudyRoomDTO[], total: number}> => {
    const response = await api.get('/study-rooms/public', { params: { page, pageSize } });
    return response.data || { rooms: [], total: 0 };
  },
  
  // 获取我的自习室列表
  getMyStudyRooms: async (): Promise<StudyRoomDTO[]> => {
    const response = await api.get('/study-rooms/mine');
    return response.data || [];
  },
  
  // 获取自习室详情
  getStudyRoomDetail: async (roomId: number): Promise<StudyRoomDetailDTO> => {
    const response = await api.get(`/study-rooms/${roomId}`);
    return response.data;
  },
  
  // 创建自习室
  createStudyRoom: async (roomData: CreateStudyRoomDTO): Promise<StudyRoomDTO> => {
    const response = await api.post('/study-rooms', roomData);
    return response.data;
  },
  
  // 更新自习室
  updateStudyRoom: async (roomId: number, roomData: UpdateStudyRoomDTO): Promise<StudyRoomDTO> => {
    const response = await api.put(`/study-rooms/${roomId}`, roomData);
    return response.data;
  },
  
  // 删除自习室
  deleteStudyRoom: async (roomId: number): Promise<void> => {
    await api.delete(`/study-rooms/${roomId}`);
  },
  
  // 加入自习室
  joinStudyRoom: async (roomId: number, isAnonymous: boolean = false): Promise<void> => {
    await api.post(`/study-rooms/${roomId}/join`, { is_anonymous: isAnonymous });
  },
  
  // 通过分享链接加入自习室
  joinByShareLink: async (shareLink: string, isAnonymous: boolean = false): Promise<StudyRoomDTO> => {
    const response = await api.post(`/study-rooms/join/${shareLink}`, { is_anonymous: isAnonymous });
    return response.data;
  },
  
  // 离开自习室
  leaveStudyRoom: async (roomId: number): Promise<void> => {
    await api.post(`/study-rooms/${roomId}/leave`);
  },
  
  // 更新成员状态
  updateMemberStatus: async (roomId: number, status: string): Promise<void> => {
    await api.put(`/study-rooms/${roomId}/status`, { status });
  }
}; 